common/tools/esm4mocha.mjs

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { readFile, stat, constants } from "node:fs/promises";
+import { dirname, join } from "node:path";
+import { fileURLToPath } from "url";
+import { Project } from "./dev-tool/node_modules/ts-morph/dist/ts-morph.js";
+
+function consoleLog(...args) {
+  if (process.env.DEBUG && process.env.DEBUG.startsWith("esm4mocha")) {
+    console.log(args);
+  }
+}
+
+function consoleDir(...args) {
+  if (process.env.DEBUG && process.env.DEBUG.startsWith("esm4mocha")) {
+    console.dir(args);
+  }
+}
+
+export async function resolve(specifier, context, defaultResolve) {
+  consoleLog("resolving...");
+  consoleDir({ specifier, context });
+  const resolved = await defaultResolve(specifier, context, defaultResolve);
+  consoleDir({ resolved });
+  if (resolved.url.includes("dist-esm/")) {
+    resolved.format = "module";
+  }
+  return resolved;
+}
+
+async function updateSpecifierValueIfRelative(declaration, base) {
+  const specifierValue = declaration.getModuleSpecifierValue();
+  consoleDir({ base, specifierValue });
+  if (specifierValue?.startsWith(".")) {
+    const sourcePath = join(base, specifierValue);
+    try {
+      const s = await stat(sourcePath, constants.R_OK);
+      consoleDir({ sourcePath });
+      if (s.isDirectory()) {
+        declaration.setModuleSpecifier(`${specifierValue}/index.js`);
+      } else {
+        declaration.setModuleSpecifier(`${specifierValue}.js`);
+      }
+    } catch (ex) {
```

Comment:
What errors does this catch? Does it make sense to fail-fast in this case or do we expect errors as part of the flow and they are fine?

Reply:
Yes this is the 2nd case when the import is like `./folder/index` but the actual module is `./folder/index.js` we would get NOENT error.  The `else` case above is rare as we most likely won't have a file with that name (no extension) on disk.